---
title: "DEseq2"
author: "RuiLi, Kai Hu"
date: "05/25/2020"
output:
  html_document:
    toc: yes
    code_folding: hide
    toc_float: yes
params: 
  max_fdr: 0.05
  min_lfc: 0.585
  indfilter: FALSE
  countFile: "../meta/COUNT.txt"
  annoFile: 'https://raw.githubusercontent.com/hukai916/Collections/master/gencode.vM21.annotation.txt'
  metaFile: "../meta/meta.xlsx"
  contrastFile:  "../meta/contrast.de.xlsx"
---

## Notes:
- Please use TPM.xlsx for normalized expression values
- Please use LFC_shrinked for Log2FoldChange
- Please use FDR for significance filtering, not p-values
- Please search for "Warning" in the report for possible problems

```{r setup, include=F}
# Bioconductor
library(BiocManager)
library(DESeq2)
library(scater)
library(EnhancedVolcano)

# CRAN
library(ggrepel)
library(ggplot2)
library(ashr)
library(MASS)
library(WriteXLS)
library(plyr)
library(gdata)
library(dplyr)
library(RColorBrewer)
library(pheatmap)
library(PoiClaClu)
library(gridExtra)
library(grid)
library(corrplot)

dir.create("../DESeq2", showWarnings = FALSE)
dir.create("../DESeq2/rnk", showWarnings = FALSE)
knitr::opts_knit$set(root.dir = "../DESeq2")
```


## Params
```{r}
max_fdr <- params$max_fdr
min_lfc <- params$min_lfc
indfilter <- params$indfilter

countFile <- params$countFile
annoFile <- params$annoFile
metaFile <- params$metaFile
contrastFile <-params$contrastFile

paste("Data file name:", countFile)
paste("MetaData file name:", metaFile)
paste("Contrast file name:", contrastFile)
paste("Annotation file name:", annoFile)
paste("FDR cut-off:", max_fdr)
paste("Log2FC cut-off:", min_lfc)
```

## Functions
```{r}
readExcel <- function(fname){
    df <- readxl::read_xlsx( fname, na='NA', sheet=1)  # na term important
    df <- data.frame(df)  #important
    return (df)
}

get_density <- function(x, y, ...) {
  dens <- MASS::kde2d(x, y, ...)
  ix <- findInterval(x, dens$x)
  iy <- findInterval(y, dens$y)
  ii <- cbind(ix, iy)
  return(dens$z[ii])
}


volcanoplot <- function(res, anno, name='name', 
                        lfcthresh=2, sigthresh=0.05, 
                        labelsig=FALSE, xlim=100, ylim=1000, textcx=1) {
  # If Annotation file is wrong, No annotation match
  res0 <- res
  res<-merge(data.frame(res0), anno, by.x = 0, by.y=1, all.x=T, sort=F)
  if (sum(anno[, 1] %in% row.names(res0)) < 1) {
    warning(
      c("\nThe annotation file and the count filt does have no match in gene-id:\n", 
        "count table gene-id: ", head(row.names(res0), 1), 
        "\nanno table gene-id: ", anno[1:1, 1], 
        "gene-id rather than gene-name used for Volcano-plot"
      ))
    res$Name <- res[, 1]
  }
  
  # remove NA
  res$padj[is.na(res$padj)] <- 1
  # set lim on x, y
  res$padj[res$padj < 10^(-ylim) & !is.na(res$padj)] <- 10^(-ylim) # y-axis top value 50
  res$log2FoldChange[res$log2FoldChange > xlim] <- xlim
  res$log2FoldChange[res$log2FoldChange < -xlim] <- -xlim
  # show num_pos num_neg
  pos <- subset(res, padj<sigthresh & log2FoldChange>lfcthresh)
  neg <- subset(res, padj<sigthresh & log2FoldChange< -lfcthresh)
  pos.n <- dim(pos)[1]
  neg.n <- dim(neg)[1]
  title=paste("Volcano Plot",
              name,
              paste("Up:", pos.n, ", Down:", neg.n, sep = ""),
              sep='\n')
  
  #labelcut <- res$padj[order(res$padj)][100] 
  
  p <- EnhancedVolcano(res,
                       lab = res$Name,
                       #selectLab = as.character(res$Name[which(res$padj<labelcut)]), # mark top genes
                       #selectLab = c("FOS", "LDHA"), # mark selected genes
                       x = 'log2FoldChange',
                       y = 'padj',
                       title = title,
                       xlab = bquote(~Log[2]~ "Fold Change"),
                       ylab = bquote(~-Log[10]~italic(FDR)),
                       #xlim = c(-6, 6),
                       pCutoff = sigthresh,
                       FCcutoff = lfcthresh,
                       #pLabellingCutoff = labelcut,
                       cutoffLineType = 'twodash',
                       cutoffLineWidth = 0.8,
                       # pointSize = 1.0, # todo: fix incomparibility in R3.5 and R4
                       # labSize = 2.0,
                       # DrawConnectors = F,
                       colAlpha = 0.5,
                       legend=c("Negative",
                                paste("|LFC|>",lfcthresh,sep=""),
                                paste("FDR<",sigthresh,sep=""),
                                "Both"),
                       legendPosition = 'right',
                       legendLabSize = 10,
                       axisLabSize = 10,
                       titleLabSize = 10,
                       legendIconSize = 3.0)
  
  # plot in html
  grid.arrange(p,
               ncol=1,
               top = textGrob(' ',
                              just = c('center'),
                              gp = gpar(fontsize = 32))
  )
  grid.rect(gp=gpar(fill=NA))
  
  # save PDF
  pdf(paste(name, "pdf", sep="."), width=8, height=6)
  grid.arrange(p,
               ncol=1,
               top = textGrob(' ',
                              just = c('center'),
                              gp = gpar(fontsize = 32))
  )
  grid.rect(gp=gpar(fill=NA))
  dev.off()
}


maplot <- function (res, thresh=max_fdr, labelsig=FALSE, textcx=1, ...) {
  # todo: improve visualization
    with(res, 
         plot(baseMean, log2FoldChange, col="grey80", pch=20, cex=.5, log="x", ...))
    with(subset(res, padj<max_fdr), 
         points(baseMean, log2FoldChange, col="grey40", pch=20, cex=.5))
    if (labelsig) {
        require(calibrate)
        with(subset(res, padj<max_fdr), 
             textxy(baseMean, log2FoldChange, labs=Gene, cex=textcx, col=2))
    }
}


process_deseq_res <- function(res="lfcshrink.res", res2="results.res", name='name', anno='anno.df', norm_exp="tpm.df"){
    ## Summary
    print(name)
    print("\n>>> Summary using FDR cut-off only (LFC not used)")
    summary(res, alpha=max_fdr)
    
    print("\n>>> Summary using both FDR and LFC_shrinked cut-off")
    sig_idx <- res$padj<max_fdr & abs(res$log2FoldChange) > min_lfc
    sig_idx[is.na(sig_idx)] <- FALSE
    res_sig <- res[sig_idx,]
    print(table(sig_idx))
    
    up_idx <- res$padj<max_fdr & res$log2FoldChange > min_lfc
    up_idx[is.na(up_idx)] <- FALSE
    res_sig <- res[up_idx,]
    print(table(up_idx))
    
    down_idx <- res$padj<max_fdr & res$log2FoldChange < -min_lfc
    down_idx[is.na(down_idx)] <- FALSE
    res_sig <- res[down_idx,]
    print(table(down_idx))

    # Prep
    res.df <- as.data.frame(res)
    names(res.df)[2] <- "log2FoldChange_shrinked"
    names(res.df)[3] <- "lfcSE_shrinked"

    res2.df <- as.data.frame(res2)
    names(res2.df)[2] <- "log2FoldChange_raw"
    names(res2.df)[3] <- "lfcSE_raw"
    res2.df <- res2.df[, c(2,3)]

    resdata <- merge(res.df, res2.df, by=0, sort=F, all.x=T)
    resdata <- merge(anno, resdata, by.x=1, by.y=1, sort=F, all.y=T)
    resdata <- merge(resdata, norm_exp, by.x=1, by.y=0, all.x=T, sort=F)
    head(resdata)
    sig_idx <- resdata$padj<max_fdr & abs(resdata$log2FoldChange_shrinked) > min_lfc # important to put this line right before output sig.xlsx
    sig_idx[is.na(sig_idx)] <- FALSE
    resdata.sig <- resdata[sig_idx,]
    head(resdata.sig)

    ## Write results
    WriteXLS(x = resdata,
             ExcelFileName = paste(name, 'deseq2.xlsx', sep = '.'),
             row.names = F, SheetNames = 'sheet1', na = 'NA')  # for user

    WriteXLS(x = resdata.sig,
             ExcelFileName = paste(name, 'deseq2.sig.FDR', max_fdr,
                                   'LFC', min_lfc, 'xlsx', sep = '.'),
             row.names = F, SheetNames = 'sheet1', na = 'NA')  # for user
    ## For GSEA
    rnk <- subset(resdata, select = c("Name","log2FoldChange_shrinked"))
    colnames(rnk) <- c("# Name","log2FoldChange_shrinked")
    rnk <- rnk[order(rnk$log2FoldChange_shrinked), ]
    rnk[, 1] <- toupper(rnk[, 1])
    write.table(rnk, 
                paste('rnk/', name, '.rnk', sep = ''), 
                row.names = F, quote = F, sep='\t')


    # # Corrlelation of Length and LFC for Niraj (only after write excel, alters resdata)
    # resdata.sig.cor <- cor.test(resdata.sig$Length, 
    #                        resdata.sig$log2FoldChange_shrinked, 
    #                        method = "spearman")
    # title <- paste("Spearman Cor:", format(resdata.sig.cor$estimate, digits=2, nsmall=2),
    #                "p-value:", format(resdata.sig.cor$p.value, digits=3, nsmall=3),
    #                sep = " ")
    # 
    # resdata.sig$density <- get_density(resdata.sig$Length, resdata.sig$log2FoldChange_shrinked, n = 100)
    # # set lim (careful, only after outputting excel)
    # resdata.sig$log2FoldChange_shrinked[resdata.sig$log2FoldChange_shrinked > 10] <- 10
    # resdata.sig$log2FoldChange_shrinked[resdata.sig$log2FoldChange_shrinked < -10] <- -10
    # resdata.sig$Length[resdata.sig$Length > 20000] <- 20000
    # ggplot(resdata.sig) + 
    #     geom_point(aes(Length, log2FoldChange_shrinked, color = density)) +
    #     scale_color_viridis() +
    #     ggtitle(paste(paste("Sig-DEG for", name), title, sep = "\n") ) + 
    #     ylim(-10, 10) +        
    #     xlim(0,20000)
    
    ##  Plots
    hist(res$pvalue, breaks=50, col="grey80", # todo: improve vis
         main = paste('Histogram of p-values (un-adjusted)', name, sep = "\n"), 
         xlab = 'pvalues', ylab = 'Frequency')
    
    hist(res$padj, breaks=50, col="grey", 
         main = paste('Histogram of Q-values (adjusted)', name, sep = "\n"), 
         xlab = 'padj', ylab = 'Frequency')
    
    maplot(res, main=paste("MAplot", paste(name, "LFC_shrinked"), sep="\n")) # todo: improve vis
    maplot(res2, main=paste("MAplot", paste(name, "LFC_raw"), sep="\n"))
    
    volcanoplot(res, anno, lfcthresh=min_lfc, sigthresh=max_fdr,
                textcx=.8,  name= paste(name, "LFC_shrinked", sep="."))
    volcanoplot(res2, anno, lfcthresh=min_lfc, sigthresh=max_fdr,
                textcx=.8,name= paste(name, "LFC_raw", sep="."))

}
```


## Read Data
```{r}
df <- read.table(countFile, 
                 sep="\t", header=TRUE, 
                 row.names = 1) # row.name in cts(matrix)
# clean names
colnames(df) <- gsub("\\.bam$", "", colnames(df))
colnames(df) <- gsub("sorted_reads.", "", colnames(df))
colnames(df) <- gsub("mapped_reads.", "", colnames(df))

# get cts
cts <- df[, 6:ncol(df)]
cts <- as.matrix(cts)

# Report Summary
print(paste("lib-size in millions:"))
print(format(colSums(cts/1e6), digits=1))
print(paste("Dim of input data:"))
print(dim(cts))

```

## Filtering
```{r}
min_rowsum <- 10
expression_filter <- rowSums(cts) >= min_rowsum  # default 10
cts <- cts[expression_filter, ]
df <- df[expression_filter, ]
paste("Removed genes with less than ", min_rowsum, "reads/fragments across all samples")
paste("Data dim after filtering:")
dim(cts)
print("Head of filtered data:")
head(cts)
```

## Read Anno
```{r}
# from GitHub (must be raw, not zipped)
getAnnotation <- function(urlpath) {
  tmp <- tempfile()
  download.file(urlpath, destfile = tmp, method = 'auto')
  return(read.table(tmp, header = TRUE))
}

paste("Getting data from", annoFile)
anno <- getAnnotation(annoFile)
print("Dimention of annotation table: ")
dim(anno)
head(anno)

if (sum(anno[, 1] %in% row.names(cts)) < 1){
  warning("!!! Annotation file and count file have no ID in common")
  warning("The results will be unannotated")
  warning("Please Double check Annotation file")
  print("count table ID:")
  print(row.names(cts)[1:2])
  print("anno table ID:")
  print(anno[1:2, 1])
  }
```

## COUNT output
```{r}
count <- data.frame(cts)
colnames(count) <- paste(colnames(count),"COUNT", sep = ":")
count_out <- merge(anno, count, by.x=1, by.y=0, all.y=T, sort=F)
head(count_out)
WriteXLS(x = count_out, 
         ExcelFileName = 'COUNT.xlsx', row.names = F, SheetNames = 'sheet1', na = 'NA')
print("saved in COUNT.xlsx")
```

## TPM calculation
```{r}
tpm <- calculateTPM(cts, df$Length)
tpm <- data.frame(tpm)
colnames(tpm) <- paste(colnames(tpm),"TPM",  sep = ":")
tpm_out <- merge(anno, tpm, by.x=1, by.y=0, all.y=T, sort=F)
# head(tpm_out)
# tail(tpm_out)
WriteXLS(x = tpm_out, 
         ExcelFileName = 'TPM.xlsx', row.names = F, SheetNames = 'sheet1', na = 'NA')
print("saved in TPM.xlsx")
```

## FPKKM calculation
```{r}
fpkm <- calculateFPKM(cts, df$Length)
fpkm <- data.frame(fpkm)
colnames(fpkm) <- paste(colnames(fpkm), "FPKM", sep = ":")
fpkm_out <- merge(anno, fpkm, by.x=1, by.y=0, all.y=T, sort=F)
# head(fpkm_out)
# tail(fpkm_out)
WriteXLS(x = fpkm_out, 
         ExcelFileName = 'FPKM.xlsx', row.names = F, SheetNames = 'sheet1', na = 'NA')
print("saved in FPKM.xlsx")
print("Recommend to use TPM, rather than FPKM")
```

## Experiment Design from Meta Data
```{r}
meta.df <- readExcel(metaFile)
meta.df <- meta.df[match(colnames(cts), meta.df$SAMPLE_LABEL), ]

contrast.df <- readExcel(contrastFile)

sample <- factor(meta.df$SAMPLE_LABEL)
batch <- factor(meta.df$BATCH)
group <- factor(meta.df$GROUP_LABEL)

coldata <- data.frame(row.names=colnames(cts), 
                      sample,
                      group,
                      batch
                      )
coldata
```

## Model fitting
```{r}
if (length(levels(batch)) > 1){
  dds <- DESeqDataSetFromMatrix(countData = cts, 
                                colData = coldata, 
                                design = ~  0 + group + batch)
}else{
  dds <- DESeqDataSetFromMatrix(countData = cts, 
                                colData = coldata, 
                                design = ~  0 + group)  # converted to alph-order
}
dds
dds <-DESeq(dds)
resultsNames(dds)
#saveRDS(dds, file = 'deseq2.dds.rds')
```

## QC Plots

<!-- ## Data transformation -->
<!-- ```{r} -->
<!-- #vsd <- vst(dds, blind=FALSE) -->
<!-- rld <- rlog(dds, blind=FALSE) -->
<!-- counts <- counts(dds, normalized=0) -->
<!-- logCounts <- log10(counts +1 ) -->

<!-- normed <- counts(dds, normalized=1) -->
<!-- logNormed <- log10(normed+1) -->
<!-- ``` -->

### Histogram of Log10(Counts)
```{r}
tmp <- log10(counts(dds) + 1)
hist(tmp, main = 'Histogram of log10(count + 1)', 100) # by default, use non-normalized data by counts function
unlink(tmp)
```

### Dispersion plot
```{r}
plotDispEsts(dds, main="Dispersion plot")
```


### Sample PCA plot
```{r}
plotQC_PCA <- function(dds) {
  vsd <- varianceStabilizingTransformation(dds) # fixed num < num(rowsum>5)
  
  pcaData <- plotPCA(vsd, intgroup = 'group', returnData=TRUE) # labeling fixed
  percentVar <- round(100 * attr(pcaData, 'percentVar'))
  
  ggplot(pcaData, aes(PC1, PC2, color = group, shape = batch)) +
  geom_point(size = 3) +
  xlab(paste0("PC1: ", percentVar[1], "% variance")) +
  ylab(paste0("PC2: ", percentVar[2], "% variance")) +
  geom_label_repel(aes(label = sample),
                    box.padding = 0.35,
                    point.padding = 1,
                    segment.color = 'grey50',
                    show.legend = FALSE) + # if TRUE, legend display might not be correct
  theme_classic()
}

plotQC_PCA(dds)
```




### Sample Heatmap (Poisson Distance Based)
```{r}
colors <- colorRampPalette( rev(brewer.pal(9, "Blues")) )(255)
poisd <- PoissonDistance(t(counts(dds)))
samplePoisDistMatrix <- as.matrix( poisd$dd ) 
rownames(samplePoisDistMatrix) <- coldata$sample
colnames(samplePoisDistMatrix) <- NULL 
colors <- colorRampPalette( rev(brewer.pal(9, "Blues")) )(255)
pheatmap(samplePoisDistMatrix,
         clustering_distance_rows=poisd$dd,
         clustering_distance_cols=poisd$dd,
         col=colors, 
         clustering_method='complete'
        )
```

### Sample Heatmap (Spearman Corr Based)
```{r}
corr <- cor(counts(dds), method="spearman")
#corr <- round(corr, 2)
corrplot(corr, 
         method="color", type="upper", add=F,  # add-on to previous plot
         is.corr=F,  # flexible coloring if False
         order="hclust", hclust.method="complete",
         tl.pos="n"
)
corrplot(corr, 
         method="number", type="lower", add=T,  # add-on to previous plot
         is.corr=T,  # flexible coloring if False
         order="hclust", hclust.method="complete",
         tl.pos="lt"
)
```

## DE analysis
```{r, warning=T}
gsub("group", "", resultsNames(dds))

for (i in 1:dim(contrast.df)[2]){
  name1 <- contrast.df[1,i]
  name2 <- contrast.df[2,i]
  name1 <- gsub(" ", "", name1)
  name2 <- gsub(" ", "", name2)
  name1 <- gsub(";$", "", name1)
  name2 <- gsub(";$", "", name2)
  name1s <- strsplit(name1, ";") [[1]]
  name2s <- strsplit(name2, ";") [[1]]
  name1 <- gsub(";", ".", name1)
  name2 <- gsub(";", ".", name2)
  
  name <- paste(name1, name2, sep = "_vs_")
  print(paste(">>>", i, name))
  
  # contrast <- c("group", name1, name2)
  
  poss <- match(name1s, gsub("group", "", resultsNames(dds)))
  negs <- match(name2s, gsub("group", "", resultsNames(dds)))
  contrast <- rep(0, length(resultsNames(dds)))
  contrast[poss] <- 1/length(poss)
  contrast[negs] <- -1/length(negs)
  print(data.frame(resNames=gsub("group", "", resultsNames(dds)), 
                   contrast=contrast))
  
  res <- lfcShrink(dds, contrast = contrast, type = 'ashr')
  res2 <- results(dds, contrast = contrast, independentFilter=indfilter)
  process_deseq_res(res = res, res2=res2, name = name, anno = anno, norm_exp = tpm)
}
```

## Log
```{r}
save.image(file="../DESeq2/RSession")
sessionInfo()
```

